# Part1

In [158]:
import re


lines = open("example1.txt").read()
lines = open("input.txt").read()

lines = lines.split("\n")

# ############################################################################
# print(lines[10:])

dish = [[c for c in line] for line in lines]
dish[:10]

[['O',
  '.',
  '.',
  '#',
  'O',
  'O',
  '.',
  '.',
  '#',
  '.',
  '.',
  '.',
  'O',
  '.',
  '.',
  'O',
  '#',
  'O',
  '.',
  'O',
  '.',
  '.',
  '.',
  'O',
  '#',
  '#',
  '.',
  'O',
  '#',
  '.',
  '.',
  '#',
  'O',
  '#',
  '.',
  '.',
  '.',
  'O',
  '.',
  'O',
  '.',
  '.',
  '.',
  '.',
  '.',
  'O',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  'O',
  '.',
  '.',
  '#',
  '.',
  '.',
  '.',
  '#',
  '.',
  '.',
  '.',
  '.',
  '.',
  'O',
  '.',
  'O',
  '.',
  'O',
  '.',
  'O',
  '#',
  '.',
  '.',
  'O',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  'O',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.'],
 ['.',
  '.',
  '.',
  '.',
  '.',
  'O',
  '#',
  '.',
  '#',
  'O',
  '.',
  '.',
  'O',
  'O',
  '.',
  '#',
  '.',
  '.',
  '.',
  '.',
  '.',
  '.',
  'O',
  '.',
  'O',
  '.',
  '.',
  '#',
  '.',
  '.',
  '.',
  '#',
  '.',
  '.',
  'O',
  'O',
  '.',
  '.',
  'O',
  '.',
  '.',
  '#',
  '#'

In [107]:
def transpose(matrix):
    # Transpose the 2D list
    return list(row for row in zip(*matrix))


In [113]:
# transpose
transposed = transpose(dish)
_ = [print(line) for line in transposed]

('O', 'O', '.', 'O', '.', 'O', '.', '.', '#', '#')
('.', '.', '.', 'O', 'O', '.', '.', '.', '.', 'O')
('.', 'O', '.', '.', '.', '#', 'O', '.', '.', 'O')
('.', 'O', '.', '#', '.', '.', '.', '.', '.', '.')
('.', '#', '.', 'O', '.', '.', '.', '.', '.', '.')
('#', '.', '#', '.', '.', 'O', '#', '.', '#', '#')
('.', '.', '#', '.', '.', '.', 'O', '.', '#', '.')
('.', '.', '.', '.', 'O', '#', '.', 'O', '#', '.')
('.', '.', '.', '.', '#', '.', '.', '.', '.', '.')
('.', '#', '.', 'O', '.', '#', 'O', '.', '.', '.')


In [150]:
line = dish[3]

print("".join(line))

def move_rocks_left(line):
    chunks = "".join(line).split("#")
    new_chunks = []
    new_chunks = ["".join(sorted(chunk)[::-1]) for chunk in chunks]
    new_chunks = "#".join(new_chunks)
    return [nc for nc in new_chunks]

new_line = move_rocks_left(line)

OO.#O....O


In [153]:
# Tranpose once
transposed = transpose(dish)

# Move rock to the left
_ = [print(t) for t in dish]
for i, line in enumerate(transposed):
    transposed[i] = move_rocks_left(line)

# Tranpose the matrix back, and its like the rocks moved up
transposed = transpose(transposed)
transposed = transpose(transposed)
transposed = transpose(transposed)

print()
_ = [print("".join(t)) for t in transposed]

['O', '.', '.', '.', '.', '#', '.', '.', '.', '.']
['O', '.', 'O', 'O', '#', '.', '.', '.', '.', '#']
['.', '.', '.', '.', '.', '#', '#', '.', '.', '.']
['O', 'O', '.', '#', 'O', '.', '.', '.', '.', 'O']
['.', 'O', '.', '.', '.', '.', '.', 'O', '#', '.']
['O', '.', '#', '.', '.', 'O', '.', '#', '.', '#']
['.', '.', 'O', '.', '.', '#', 'O', '.', '.', 'O']
['.', '.', '.', '.', '.', '.', '.', 'O', '.', '.']
['#', '.', '.', '.', '.', '#', '#', '#', '.', '.']
['#', 'O', 'O', '.', '.', '#', '.', '.', '.', '.']

OOOO.#.O..
OO..#....#
OO..O##..O
O..#.OO...
........#.
..#....#.#
..O..#.O.O
..O.......
#....###..
#....#....


### Main code

In [160]:
transposed = transpose(dish)

for i, line in enumerate(transposed):
    transposed[i] = move_rocks_left(line)

transposed = transpose(transposed)
transposed = transpose(transposed)
transposed = transpose(transposed)


num_rows = len(transposed)
SOLUTION = 0
for i, line in enumerate(transposed):
    SOLUTION += line.count("O") * (num_rows - i)

print(f"SOLUTION: {SOLUTION}")

SOLUTION: 106648


##### Solution: 106.648 (106648)

<br>

---

# Part 2
The rocks roll **north**, then **west**, then **south**, then **east**

1.000.000.000 times!!!!  (1000000000)

In [161]:
import re


lines = open("input.txt").read()
lines = open("example1.txt").read()

lines = lines.split("\n")

# ############################################################################
# print(lines[10:])

dish = [[c for c in line] for line in lines]
dish[:10]

[['O', '.', '.', '.', '.', '#', '.', '.', '.', '.'],
 ['O', '.', 'O', 'O', '#', '.', '.', '.', '.', '#'],
 ['.', '.', '.', '.', '.', '#', '#', '.', '.', '.'],
 ['O', 'O', '.', '#', 'O', '.', '.', '.', '.', 'O'],
 ['.', 'O', '.', '.', '.', '.', '.', 'O', '#', '.'],
 ['O', '.', '#', '.', '.', 'O', '.', '#', '.', '#'],
 ['.', '.', 'O', '.', '.', '#', 'O', '.', '.', 'O'],
 ['.', '.', '.', '.', '.', '.', '.', 'O', '.', '.'],
 ['#', '.', '.', '.', '.', '#', '#', '#', '.', '.'],
 ['#', 'O', 'O', '.', '.', '#', '.', '.', '.', '.']]

In [169]:
line = dish[3]

print("".join(line))

def move_rocks(line, invert: bool = False):
    chunks = "".join(line).split("#")
    new_chunks = []
    if (invert):
        new_chunks = ["".join(sorted(chunk)[::-1]) for chunk in chunks]
    else:
        new_chunks = ["".join(sorted(chunk)) for chunk in chunks]
    new_chunks = "#".join(new_chunks)
    return [nc for nc in new_chunks]


new_line = move_rocks(line, True)
print("".join(new_line))

OO.#O....O
OO.#OO....


##### Solution: XXXX